src/advanced-concepts/async-programming.md

```diff
@@ -1 +1,422 @@
+>  
+> _[Assertions](/common-programming-concepts/assertions.html) also generate a trap if their argument evaluates to `false`_
+
+## Async* and Await*
+
+Recall that 'ordinary' [shared functions](/internet-computer-programming-concepts/actors.html#public-shared-functions-in-actors) with `async` return type are part of the [_actor type_](/internet-computer-programming-concepts/actors.html#actor-type) and therefore publicly visible in the [_public API_](/internet-computer-programming-concepts/async-data/candid.html#actor-interfaces) of the actor. Shared functions provide an [asynchronous context](#messaging-restrictions) from which other shared functions can be called and awaited, because awaiting a shared function requires a [message](#messages-and-atomicity) to be sent.
+
+**Private non-shared functions with `async*` return type** provide an [asynchronous context](#messaging-restrictions) without exposing the function as part of the public API of the actor.
+
+A call to an `async*` function also immediately returns a future. The future **needs** to be awaited using the `await*` keyword (in any asynchronous context `async` or `async*`) to produce a result. This was not the case with un-awaited `async` calls, see [atomic functions that send messages](#atomic-functions-that-send-messages).
+
+For demonstration purposes, lets look at an example of a private `async*` function, that does not use its asynchronous context to call other `async` or `async*` functions, but instead only performs a _delayed computation_:
+
+```motoko
+{{#include _mo/async-calls7.mo}}
+```
+
+`compute` is a [private function](/common-programming-concepts/functions.html#private-functions) with `async* Nat` return type. Calling it directly yields a _future_ of type `async* Nat` and resumes execution without blocking. This future needs to be awaited using `await*` for the computation to actually execute (unlike the case with 'ordinary' `async` [atomic functions that send messages](#atomic-functions-that-send-messages)).
+
+`await*` also suspends execution, until a result is obtained. We could also pass around the future within our actor code and only `await*` it when we actually need the result.
+
+We `await*` our function `compute` from within an [asynchronous context](#messaging-restrictions).
+
+We `await*` our function `compute` from within an 'ordinary' shared `async` function `call_compute` or from within another private `async*` like `call_compute2`.
+
+In the case of `call_compute` we obtain the result by first declaring a future and then `await*`ing it. In the case of `call_compute2` we `await*` the result directly.
+
+`compute` and `call_compute` are not part of the [_actor type_](/internet-computer-programming-concepts/actors.html#actor-type) and [_public API_](/internet-computer-programming-concepts/async-data/candid.html#actor-interfaces), because it is a [private function](/common-programming-concepts/functions.html#private-functions).
```

Comment:
```suggestion
`compute` and `call_compute` are not part of the [_actor type_](/internet-computer-programming-concepts/actors.html#actor-type) and [_public API_](/internet-computer-programming-concepts/async-data/candid.html#actor-interfaces), because they are [private functions](/common-programming-concepts/functions.html#private-functions).
```